pkg/multiplexer/headers.go

```diff
@@ -62,3 +63,35 @@ func forwardedForFromRequest(req *http.Request) string {
 
 	return host
 }
+
+type headerResponseWriter struct {
+	headers http.Header
+}
+
+func (hrw *headerResponseWriter) Header() http.Header {
+	return hrw.headers
+}
+
+func (hrw *headerResponseWriter) Write(b []byte) (int, error) {
+	return 0, errors.New("Write not supported on response")
```

Comment:
I like that we were able to re-use the same interface for the middleware, but it's unfortunate that we don't support the other methods provided by an `http.ResponseWriter`. It also makes sense though, since we don't want to allow writing the response. Since we don't support the full `http.ResponseWriter` interface and `ServeHTTP` isn't handling a request, maybe we could create our own abstraction specifically for headers.

What are your thoughts on a header specific API, maybe (very roughly) something like the following?

```go
type HeaderModifier = func(headers http.Header, results []*multiplexer.Result, r *http.Request)

for _, modifier := range server.HeaderModifiers {
  modifier(responseBuilder.Header, responseBuilder.results, r) // each modifier can mutate the http.Header value
}
```

That's just a rough idea, but it would allow us to avoid passing a `http.ResponseWriter` and having to call `ServeHTTP`. The immediate drawback is that we lose the ability to perform work after the next middleware was called, but we could also design an API that allows that, if we think it's useful.

I think the advantage of our own API is that we can provide types/interfaces that turn runtime errors like calling `Write` into compile time errors since we can avoid defining that method. Thoughts?

Reply:
> What are your thoughts on a header specific API, maybe (very roughly) something like the following?

Yeah, I waffled on this, my early version locally went down that path but after staring at it it felt really close to `http.Handler` but without the results. And since we already had the route and params in the context, re-using the request context for the results seemed consistent and then `http.Handler ` felt right.

I agree having a `http.ResponseWriter` that only supported headers is a little odd but re-using `http.Handler` allows consumers to easily wrap/customize header setting in the same way where they can wrap before caling `next` or skip calling `next` altogether.